proxy/reverse.go

```diff
@@ -203,3 +210,39 @@ func maybeSetForwardedFor(req *http.Request) {
 	}
 	req.Header.Set("X-Forwarded-For", clientIP)
 }
+
+// checkProxyLoop returns an error when it detects a proxy loop.
+func checkProxyLoop(req *http.Request, proxyID string, hopLimit int) error {
+	// check if proxyID has been seen before
+	seen := false
+	if v, ok := req.Header["X-Etcd-Proxy-Hop-Seen-By"]; ok {
+		for _, id := range strings.Split(strings.Join(v, ", "), ", ") {
+			if id == proxyID {
+				seen = true
+				break
+			}
+		}
+		// fold multiple proxyIDs into one header
+		proxyID = strings.Join(v, ", ") + ", " + proxyID
+	}
+	req.Header.Set("X-Etcd-Proxy-Hop-Seen-By", proxyID)
```

Comment:
I kept this one header `X-Etcd-Proxy-Hop-Seen-By` to record proxyIDs and detect proxy loop by checking this header in the following requests. And have a separate `struct` member `proxyURLs` to record proxy URLs in general. Once it finds out the proxy loop, it will remove the proxy URLs from the `director`'s endpoints. Thanks,